notebooks/autodiff_cookbook.ipynb

```diff
@@ -0,0 +1,1062 @@
+        "\n",
+        "If we don't commit to one specific input point $x$, then we can think of the function $\\partial f$ as first taking an input point and returning the Jacobian linear map at that input point:\n",
+        "\n",
+        "$\\qquad \\partial f : \\mathbb{R}^n \\to \\mathbb{R}^n \\to \\mathbb{R}^m$.\n",
+        "\n",
+        "In particular, we can uncurry things so that given input point $x \\in \\mathbb{R}^n$ and a tangent vector $v \\in \\mathbb{R}^n$, we get back an output tangent vector in $\\mathbb{R}^m$. We call that mapping, from $(x, v)$ pairs to output tangent vectors, the *Jacobian-vector product*, and write it as\n",
+        "\n",
+        "$\\qquad (x, v) \\mapsto \\partial f(x) v$\n",
+        "\n",
+        "#### JVPs in JAX code\n",
+        "\n",
+        "Back in Python code, JAX's `jvp` function models this transformation. Given a Python function that evaluates $f$, JAX's `jvp` is a way to get a Python function for evaluating $(x, v) \\mapsto (f(x), \\partial f(x) v)$."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "pTncYR6F6vvG",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import jvp\n",
+        "\n",
+        "# Isolate the function from the weight matrix to the predictions\n",
+        "f = lambda W: predict(W, b, inputs)\n",
+        "\n",
+        "key, subkey = random.split(key)\n",
+        "v = random.normal(subkey, W.shape)\n",
+        "\n",
+        "# Push forward the vector `v` along `f` evaluated at `W`\n",
+        "y, u = jvp(f, (W,), (v,))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "m1VJgJYQGfCK",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "In terms of Haskell-like type signatures, we could write\n",
+        "\n",
+        "```haskell\n",
+        "jvp :: (a -> b) -> a -> T a -> (b, T b)\n",
+        "```\n",
+        "\n",
+        "where we use `T a` to denote the type of the tangent space for `a`. In words, `jvp` takes as arguments a function of type `a -> b`, a value of type `a`, and a tangent vector value of type `T a`. It gives back a pair consisting of a value of type `b` and an output tangent vector of type `T b`."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "3RpbiasHGD3X",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Memory cost doesn't scale with the depth of the computation. FLOP cost about 2x evaluating just the function, with the cost of $(x, v) \\mapsto (x, \\partial f(x) v)$ has about the same cost to evaluate as $f$.\n",
+        "\n",
+        "But think of building a Jacobian matrix: one column at a time. Efficient for \"tall\" Jacobians but inefficient for \"wide\" Jacobians.\n",
```

Comment:
Remove the colon after "matrix".

Reply:
"Efficient for ..." is a sentence fragment. Rephrase.

Reply:
Done.